include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+
+  /**
+   * Calls from the VM>
+   * This collection represents all the calls into the VM in the ABI
+   * (https://github.com/proxy-wasm/spec).  They are not broken into separate interfaces because
+   * doing so would result in duplication since calls into the VM must be overriden by the
+   * implementations in the ContextBase.
+   *
+   * Note: most of these return a WasmResult, assume:
+   * @return a WasmResult with the status of the call.
+   */
+
+  /**
+   * Log a message.  Note: log_prefix() is *not* auto-prepended end may be useful to provide some
+   * context.
+   */
+  virtual WasmResult log(uint64_t /* level */, string_view /* message */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Enables a periodic timer with the given period or sets the period of an existing timer. Note:
+   * the timer is associated with the Root Context of whatever Context this call was made on and
+   * there is only one timer available per Root Context.
+   * @param period is the period of the periodic timer in milliseconds.
+   */
+  virtual WasmResult setTimerPeriod(std::chrono::milliseconds /* period */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Provides the current time in nanoseconds.
+   */
+  virtual uint64_t getCurrentTimeNanoseconds() {
+    struct timespec tpe;
+    clock_gettime(CLOCK_REALTIME, &tpe);
+    uint64_t t = tpe.tv_sec;
+    t *= 1000000000;
+    t += tpe.tv_nsec;
+    return t;
+  }
+
+  /**
+   * Provides the status of the last call into the VM or out of the VM, similar to errno.
+   * @return the status code and a descriptive string.
+   */
+  virtual std::pair<uint32_t, string_view> getStatus() {
+    unimplemented();
+    return std::make_pair(1, "unimplemented");
+  }
+
+  // Buffer
+
+  /**
+   * Provides a BufferInterface to be used to return buffered data to the VM.
+   * @param type is the type of buffer to provide.
+   */
+  virtual const BufferInterface *getBuffer(WasmBufferType /* type */) {
+    unimplemented();
+    return nullptr;
+  }
+
+  /**
+   * Provides the end-of-stream status of the current stream (if any) or false.
+   */
+  virtual bool end_of_stream() {
+    unimplemented();
+    return true;
+  }
+
+  // HTTP
+
+  /**
+   * Make an outgoing HTTP request.
+   * @param target specifies the proxy-specific target of the call (e.g. a cluster, peer, or host).
+   * @param request_headers are the request headers.
+   * @param request_body is the request body.
+   * @param request_trailers are the request trailers.
+   * @param timeout_milliseconds is a timeout after which the request will be considered to have
+   * failed.
+   * @param token_ptr contains a pointer to a location to store the token which will be used with
+   * the corresponding onHttpCallResponse.
+   */
+  virtual WasmResult httpCall(string_view /* target */, const Pairs & /*request_headers */,
```

Comment:
I think, `httpCall()` method should go into a dedicated `HttpClient` interface, group of `grpc*()` methods - into `GrpcClient` interface, group of `*Metric()` methods - into `Metrics` interface, group of `*SharedData()` - into `SharedData` interface, group of `*SharedQueue()` methods - into `SharedQueues` interface, etc.

Envoy will always be providing more services for extensions to use than a standard that aims to suit all proxies. That is no different from browsers.

As an Envoy contributor, I wouldn't want to start development of every new Wasm feature by making changes to the standard.

So, API should be optimised for a scenario where Envoy is ahead of the standard.

Reply:
+1 this is what I had in mind also. It will vastly help with readability even if we continue to use the "single context rules them all" approach.

Reply:
+1, but to @yskopets point, I think we need to balance API stability against Envoy developer productivity. We need some way to keep growing/pushing the API forward, without breaking backward compatibility. I almost wish we had done these API definitions in proto and used something like protolock.

Reply:
We could use a syscall approach: one call to rule them all.  I wouldn't use protos though as protos are very slow, particularly in Wasm for small values (3x+ slower than native code).  We could use something like flatbuffers. but we need to consider different language support.  The current ABI format is built in part to make it easier to add additional languages where the lowest common denominator is often a "C" call with primitive values.

Reply:
I can't speak for @htuch but I think he was expressing a desire to describe the API in a structured way whether that is swagger/proto/etc. IMO that's aspirational. :)

Reply:
(Aspirational at least for the MVP)